server.py

```diff
@@ -1,36 +1,376 @@
+        Data to match on.
+    :return:
+        A list of tuples containing the match score and the matched coach.
+    """
+
+    def get_lang_match_score(
+        c: Coach, languages: Dict[str, int]
+    ) -> Optional[Tuple[str, int]]:
+        match_langs = [L for L in languages if L in c.languages]
+        return max(
+            (0, *[30 - 2 * (languages[L] * c.languages[L]) for L in match_langs])
+        )
+
+    coach_matches = []
+    for coach in get_coaches():
+        if not coach.available:
+            continue
+        match_score = 0
+        match_score += get_lang_match_score(coach, data.get("languages", []))
+        if data.get("need") in coach.need:
+            match_score += 20
+        if data.get("rights") in coach.rights:
+            match_score += 10
+        if data.get("housing") in coach.housing:
+            match_score += 10
+        if data.get("gender") == coach.gender:
+            match_score += 5
+        if "birth_year" in data:
+            match_score += max(0, 10 - abs(data["birth_year"] - coach.birth_year))
+        coach_matches.append((match_score, coach))
+
+    return sorted(coach_matches, reverse=True)
+
+
+# ------------------------------------------------------------------------------
+# Routes
+# ------------------------------------------------------------------------------
+
+
+COMMON_HEADERS = {"Access-Control-Allow-Origin": "http://localhost:3000"}
 
-    return decorated_function
 
 @app.route("/")
-@authorized
-async def index(request):
-  return await sanic.response.file("build/index.html")
+async def index(request: sanic.request.Request) -> sanic.response.HTTPResponse:
+    return await sanic.response.file("build/index.html")
+
+
+@app.route("/api/v1/coaches", methods=["GET", "POST"])
+@app.route("/api/v1/coaches/<coach_id:int>", methods=["GET", "POST", "DELETE"])
+async def api_coaches(
+    request: sanic.request.Request, coach_id: Optional[int] = None
+) -> sanic.response.HTTPResponse:
+    """HTTP API for 'coaches'."""
+    try:
+        if request.method == "GET":
+            if coach_id is None:
+                coaches = get_coaches()
+                response = sanic.response.json([c.to_json() for c in coaches])
+            else:
+                coach = get_coach(coach_id)
+                response = sanic.response.json(coach.to_json())
+        elif request.method == "POST":
+            if coach_id is None:
+                coach = create_coach(request.json)
+            else:
+                coach = edit_coach(coach_id, request.json)
+            response = sanic.response.json(coach.to_json())
+        elif request.method == "DELETE":
+            delete_coach(coach_id)
+            response = sanic.response.empty()
+        else:
+            assert False
+    except Exception:
+        logger.exception(
+            "Unexpected exception on %s route", request.path, exc_info=True
+        )
+        response = sanic.response.text("Error processing request", status=400)
```

Comment:
Are there any errors which are known to bubble up to this point? At the least I'd expect some error information to be sent back to the client - is there a good way to do this?

Reply:
I think the only expected errors would be through misuse of the HTTP API, e.g. missing/unexpected format of URL args. But yes, error information should definitely be returned, I just wasn't sure of the best way to do so (so did nothing for now!).

Options:
 - Return an error message in the text
   - Either just the string form of the exception or something more suitable for user consumption
 - Return JSON (complicates the API sightly but allows for adding fields, i.e. errors are structured data)
   - E.g. Could return `sanic.response.json({"error": <error string>})`

If we want the errors returned to be suitable for end-user-consumption then this would all need a bit of thought with catching and reraising exceptions all over the place. I've done this kind of thing a few times before (and I think it's often the 'right' thing to do) but I find it a bit unpleasant and fiddly. I'd ask to defer this if this is the choice, so maybe the question should really be what to do in the short term... (which is how I arrived at the current solution!)

Thoughts?

Reply:
I think the current solution is fine for the short-term, since catching and re-raising exceptions seems preferable to allowing any of them to bubble up - but will take longer!

WRT the format of the error message returned, I think JSON could be preferable since then you could have e.g. the traceback, a user-friendly error string, and the type of exception raised (for example)?

I'll merge in for now, and we should create a Github issue for improving the way errors are returned!

Reply:
Sounds good. Raised https://github.com/richcooper95/crisis/issues/9.